sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala

```diff
@@ -0,0 +1,98 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.ExpressionBuilder
+import org.apache.spark.sql.catalyst.expressions.codegen._
+import org.apache.spark.sql.catalyst.util.CollationFactory
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.types._
+
+@ExpressionDescription(
+  usage = "_FUNC_(expr, collationName)",
+  examples = """
+    Examples:
+      > SELECT COLLATION('Spark SQL' _FUNC_ 'UCS_BASIC_LCASE');
+       UCS_BASIC_LCASE
+  """,
+  since = "4.0.0",
+  group = "string_funcs")
+object CollateExpressionBuilder extends ExpressionBuilder {
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    expressions match {
+      case Seq(e: Expression, collationExpr: Expression) =>
+        (collationExpr.dataType, collationExpr.foldable) match {
+          case (StringType, true) =>
+            val evalCollation = collationExpr.eval()
+            if (evalCollation == null) {
+              throw QueryCompilationErrors.unexpectedNullError("collation", collationExpr)
+            } else {
+              Collate(e, collationExpr.eval().toString)
+            }
+          case (StringType, false) => throw QueryCompilationErrors.nonFoldableArgumentError(
+            funcName, "collationName", StringType)
+          case (_, _) => throw QueryCompilationErrors.unexpectedInputDataTypeError(
+            funcName, 1, StringType, collationExpr)
+        }
+      case s => throw QueryCompilationErrors.wrongNumArgsError(funcName, Seq(2), s.length)
+    }
+  }
+}
+
+/**
+ * An expression that marks a given expression with specified collation.
+ * This function is pass-through, it will not modify the input data.
+ * Only type metadata will be updated.
+ */
+case class Collate(child: Expression, collationName: String)
+  extends UnaryExpression with ExpectsInputTypes {
+  private val collationId = CollationFactory.collationNameToId(collationName)
+  override def dataType: DataType = StringType(collationId)
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType)
+
+  override protected def withNewChildInternal(
+    newChild: Expression): Expression = copy(newChild)
+
+  override def eval(row: InternalRow): Any = child.eval(row)
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
+    defineCodeGen(ctx, ev, (in) => in)
+}
+
+/**
+ * An expression that returns the collation name of a given expression.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(expr)",
```

Comment:
Please, describe it. You can move the comment above to here.

Reply:
done!